controllers/apply.go

```diff
@@ -69,19 +101,170 @@ func getK8sSensorDeployments(
+
+	// Return true if targets are different (update needed)
+	return currentTargets != newTargets
+}
+
+// CreateDeploymentContext creates a deployment context for the k8s-sensor deployment.
+// It handles both OpenShift and vanilla Kubernetes cases, setting up the appropriate
+// ETCD configuration based on the environment.
+func CreateDeploymentContext(
+	ctx context.Context,
+	deps DeploymentContextDependencies,
+	agent *instanav1.InstanaAgent,
+	isOpenShift bool,
+	logger logr.Logger,
+) (*k8ssensordeployment.DeploymentContext, error) {
+	var deploymentContext *k8ssensordeployment.DeploymentContext
+
+	// For OpenShift, create the service-CA ConfigMap
+	if isOpenShift {
+		if err := deps.CreateServiceCAConfigMap(ctx, agent, logger); err != nil {
+			logger.Error(err, "Failed to create service-CA ConfigMap")
+			// Continue with reconciliation, don't fail the whole process
+			return deploymentContext, nil
+		}
+
+		// Set up deployment context for OpenShift
+		deploymentContext = &k8ssensordeployment.DeploymentContext{
+			ETCDCASecretName: constants.ServiceCAConfigMapName,
+		}
+		return deploymentContext, nil
+	}
+
+	// For vanilla Kubernetes, discover ETCD endpoints
+	discoveredETCD, err := deps.DiscoverETCDEndpoints(ctx, agent, logger)
+	if err != nil {
+		logger.Error(err, "Failed to discover ETCD endpoints")
+		// Continue with reconciliation, don't fail the whole process
+		return deploymentContext, nil
+	}
+
+	if discoveredETCD == nil || len(discoveredETCD.Targets) == 0 {
+		return deploymentContext, nil
+	}
+
+	// Check if we need to update the Deployment with new ETCD targets
+	existingDeployment := &appsv1.Deployment{}
+	helperInstance := helpers.NewHelpers(agent)
+	err = deps.Get(ctx, client.ObjectKey{
+		Namespace: agent.Namespace,
+		Name:      helperInstance.K8sSensorResourcesName(),
+	}, existingDeployment)
+
+	if err != nil {
+		logger.Info("K8sensor deployment not found, will create with discovered ETCD targets")
+	} else {
+		// Compare current and discovered ETCD targets
+		needsUpdate := compareAndUpdateETCDTargets(existingDeployment, discoveredETCD.Targets, logger)
+		if !needsUpdate {
+			logger.Info("ETCD targets unchanged, skipping Deployment update")
+			return nil, nil
+		}
+	}
+
+	// Use sorted targets for consistency
+	sortedTargets := getSortedTargets(discoveredETCD.Targets)
+
+	logger.Info("Using discovered ETCD targets", "targets", sortedTargets)
+	deploymentContext = &k8ssensordeployment.DeploymentContext{
+		DiscoveredETCDTargets: sortedTargets,
+	}
+	if discoveredETCD.CAFound {
+		deploymentContext.ETCDCASecretName = constants.ETCDCASecretName
+	}
+
+	return deploymentContext, nil
+}
+
+// createDeploymentContext creates a deployment context for the k8s-sensor deployment.
+// It handles both OpenShift and vanilla Kubernetes cases, setting up the appropriate
+// ETCD configuration based on the environment.
+func (r *InstanaAgentReconciler) createDeploymentContext(
```

Comment:
and make this a function... it's fine I'll approve this I think it's easier if I make this myself